kube-runtime/src/finalizer.rs

```diff
@@ -0,0 +1,146 @@
+    ApplyFailed { source: ApplyErr },
+    #[snafu(display("failed to clean up object: {}", source))]
+    CleanupFailed { source: CleanupErr },
+    #[snafu(display("failed to add finalizer: {}", source))]
+    AddFinalizer { source: kube::Error },
+    #[snafu(display("failed to remove finalizer: {}", source))]
+    RemoveFinalizer { source: kube::Error },
+}
+
+/// Reconcile an object in a way that requires cleanup before an object can be deleted. It does this by
+/// managing a [`ObjectMeta::finalizers`] entry, which prevents the object from being deleted before the
+/// cleanup is done.
+///
+/// In typical usage, if you use `finalizer` then it should be the only top-level "action"
+/// in your [`applier`]/[`Controller`]'s `reconcile` function.
+///
+/// # Expected Flow
+///
+/// 1. User creates object
+/// 2. Reconciler sees object
+/// 3. `finalizer` adds `finalizer_name` to [`ObjectMeta::finalizers`]
+/// 4. Reconciler sees updated object
+/// 5. `finalizer` runs `apply`
+/// 6. User updates object
+/// 7. Reconciler sees updated object
+/// 8. `finalizer` runs `apply`
+/// 9. User deletes object
+/// 10. Reconciler sees deleting object
+/// 11. `finalizer` runs `cleanup`
+/// 12. `finalizer` removes `finalizer_name` from [`ObjectMeta::finalizers`]
+/// 13. Kubernetes sees that all [`ObjectMeta::finalizers`] are gone and finally deletes the object
+///
+/// # Guarantees
+///
+/// If `apply` is ever started then `cleanup` must succeed before the Kubernetes object deletion completes.
+///
+/// # Assumptions
+///
+/// `finalizer_name` must be unique among the controllers interacting with the object
+///
+/// `apply` and `cleanup` must both be idempotent, and tolerate being executed several times (even if previously cancelled).
+///
+/// `cleanup` must tolerate `apply` never having ran at all, or never having succeeded. Keep in mind that
+/// even infallible `.await`s are cancellation points.
+///
+/// # Caveats
+///
+/// Object deletes will get stuck while the controller is not running, or if `cleanup` fails for some reason.
+///
+/// # Errors
+///
+/// `apply` and `cleanup` are both fallible, their errors are passed through as [`Error::ApplyFailed`]
+/// and [`Error::CleanupFailed`], respectively.
+///
+/// In addition, adding and removing the finalizer itself may fail. In particular, this may be because of
+/// network errors, or because another `finalizer` was updated in the meantime on the same object.
+pub async fn finalizer<K, ApplyFut, CleanupFut>(
+    api: &Api<K>,
+    finalizer_name: &str,
+    obj: K,
+    apply: impl FnOnce(K) -> ApplyFut,
+    cleanup: impl FnOnce(K) -> CleanupFut,
+) -> Result<ReconcilerAction, Error<ApplyFut::Error, CleanupFut::Error>>
+where
+    K: Metadata<Ty = ObjectMeta> + Clone + DeserializeOwned + Serialize + Debug,
+    ApplyFut: TryFuture<Ok = ReconcilerAction>,
+    ApplyFut::Error: StdError + 'static,
+    CleanupFut: TryFuture<Ok = ()>,
+    CleanupFut::Error: StdError + 'static,
```

Comment:
All of this error handling is slightly awkward, maybe it would be nicer to just past `apply'/`cleanup`'s return value through as-is? Then again...

1. Makes it harder to differentiate between `apply` and `cleanup` cases
2. It'd (likely) cause a resource leak if we remove the finalizer despite `cleanup` failing
3. `ReconcilerAction` isn't very useful for `cleanup`